modules/dchain.js

```diff
@@ -0,0 +1,144 @@
+    }
+  }
+
+  let dchainProps = Object.keys(dchainObj);
+  dchainProps.forEach(prop => {
+    if (!includes(dchainPropList, prop)) {
+      appendFailMsg(`dchain.${prop}` + shouldBeValid);
+    }
+  });
+
+  if (dchainObj.complete !== 0 && dchainObj.complete !== 1) {
+    appendFailMsg(`dchain.complete should be 0 or 1`);
+  }
+
+  if (!isStr(dchainObj.ver)) {
+    appendFailMsg(`dchain.ver` + shouldBeAString);
+  }
+
+  if (hasOwn(dchainObj, 'ext')) {
+    if (!isPlainObject(dchainObj.ext)) {
+      appendFailMsg(`dchain.ext` + shouldBeAnObject);
+    }
+  }
+
+  if (!isArray(dchainObj.nodes)) {
+    appendFailMsg(`dchain.nodes` + shouldBeAnArray);
+    printFailMsg();
+    if (mode === MODE.STRICT) return false;
+  } else {
+    const nodesPropList = ['asi', 'bsid', 'rid', 'name', 'domain', 'ext'];
+    dchainObj.nodes.forEach((node, index) => {
+      if (!isPlainObject(node)) {
+        appendFailMsg(`dchain.nodes[${index}]` + shouldBeAnObject);
+      } else {
+        let nodeProps = Object.keys(node);
+        nodeProps.forEach(prop => {
+          if (!includes(nodesPropList, prop)) {
+            appendFailMsg(`dchain.nodes[${index}].${prop}` + shouldBeValid);
+          }
+
+          if (prop === 'ext') {
+            if (!isPlainObject(node.ext)) {
+              appendFailMsg(`dchain.nodes[${index}].ext` + shouldBeAnObject);
+            }
+          } else {
+            if (!isStr(node[prop])) {
+              appendFailMsg(`dchain.nodes[${index}].${prop}` + shouldBeAString);
+            }
+          }
+        });
+      }
+    });
+  }
+
+  if (failMsg.length > 0) {
+    printFailMsg();
+    if (mode === MODE.STRICT) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function isValidDchain(bid) {
+  let mode = MODE.STRICT;
+  const dchainConfig = config.getConfig('dchain');
+
+  if (dchainConfig && isStr(dchainConfig.validation) && MODES.indexOf(dchainConfig.validation) != -1) {
+    mode = dchainConfig.validation;
+  }
+
+  if (mode === MODE.OFF) {
+    return true;
+  } else {
+    return checkDchainSyntax(bid, mode);
+  }
+}
+
+export function addBidResponseHook(fn, adUnitCode, bid) {
+  const basicDchain = {
+    ver: '1.0',
+    complete: 0,
+    nodes: [{ name: bid.bidderCode }]
+  };
+
+  let bidDchain = deepAccess(bid, 'meta.dchain');
+  if (bidDchain && isPlainObject(bidDchain)) {
+    let result = isValidDchain(bid);
+
+    if (result) {
+      // extra check in-case mode is OFF and there is a setup issue
+      if (isArray(bidDchain.nodes)) {
+        bid.meta.dchain.nodes.push({ asi: bid.bidderCode });
+      } else {
+        logWarn('bid.meta.dchain.nodes did not exist or was not an array; did not append prebid dchain.', bid);
+      }
+    } else {
+      // bid.meta.dchain = basicDchain;     // should we assign a backup dchain if bidder's dchain was invalid?
+      delete bid.meta.dchain; // or delete the bad object?
+    }
+  } else {
+    bid.meta.dchain = basicDchain;
+  }
+
```

Comment:
it seems like if there is no meta.dchain, but there is meta.networkid or networkname; one would expect two nodes: 

{
    ver: '1.0',
    complete: 0,
    nodes: [{ name: bid.meta.networkname, bsid: bid.meta.networkid }, { name: bid.bidderCode }]
  };

On #3115 and in doc changes to https://docs.prebid.org/dev-docs/bidder-adaptor.html these fields were standardized and adopted in some adapters eg https://github.com/prebid/Prebid.js/blob/a5141af2f738be6980f19843315e5953da501ad7/modules/pubmaticBidAdapter.js#L1257 and at https://github.com/prebid/Prebid.js/blob/a5141af2f738be6980f19843315e5953da501ad7/modules/ixBidAdapter.js#L282

Reply:
I'll put together some updates to look for these bid.meta fields and use them if they're present.  Thanks!

Reply:
Pushed a change into the PR.